aptos-move/move-examples/data_structures/sources/smart_tree.move

```diff
@@ -0,0 +1,971 @@
+                r = mid;
+            };
+        };
+
+        new_iter(leaf, l, vector::borrow(keys, l).max_key)
+    }
+
+    /// Returns an iterator pointing to the element that equals to the provided key, or an end
+    /// iterator if the key is not found.
+    public fun find<V>(tree: &SmartTree<V>, key: u64): Iterator {
+        if (!bucket_table::contains(&tree.entries, &key)) {
+            return new_end_iter(tree)
+        };
+
+        lower_bound(tree, key)
+    }
+
+    /// Returns true iff the key exists in the tree.
+    public fun contains<V>(tree: &SmartTree<V>, key: u64): bool {
+        bucket_table::contains(&tree.entries, &key)
+    }
+
+    /// Returns the key of the given iterator.
+    public fun get_key(iter: Iterator): u64 {
+        assert!(iter.node_index != NULL_INDEX, E_INVALID_PARAMETER);
+        iter.key
+    }
+
+    /// Returns a reference to the element with its key, aborts if the key is not found.
+    public fun borrow<V>(tree: &SmartTree<V>, key: u64): &V {
+        bucket_table::borrow(&tree.entries, key)
+    }
+
+    /// Returns a mutable reference to the element with its key at the given index, aborts if the key is not found.
+    public fun borrow_mut<V>(tree: &mut SmartTree<V>, key: u64): &mut V {
+        bucket_table::borrow_mut(&mut tree.entries, key)
+    }
+
+    /// Returns the number of elements in the SmartTree.
+    public fun size<V>(tree: &SmartTree<V>): u64 {
+        bucket_table::length(&tree.entries)
+    }
+
+    /// Returns true iff the SmartTree is empty.
+    public fun empty<V>(tree: &SmartTree<V>): bool {
+        bucket_table::length(&tree.entries) == 0
+    }
+
+    /// Return the begin iterator.
+    public fun new_begin_iter<V>(tree: &SmartTree<V>): Iterator {
+        if (empty(tree)) {
+            return new_iter(NULL_INDEX, 0, 0)
+        };
+
+        let node = bucket_table::borrow(&tree.nodes, tree.min_leaf_index);
+        let key = vector::borrow(&node.children, 0).max_key;
+
+        new_iter(tree.min_leaf_index, 0, key)
+    }
+
+    /// Return the end iterator.
+    public fun new_end_iter<V>(_tree: &SmartTree<V>): Iterator {
+        new_iter(NULL_INDEX, 0, 0)
+    }
+
+    /// Returns the next iterator, or none if already at the end iterator.
+    /// Requires the tree is not changed after the input iterator is generated.
+    public fun next_iter<V>(tree: &SmartTree<V>, iter: Iterator): Option<Iterator> {
+        let node_index = iter.node_index;
+        if (node_index == NULL_INDEX) {
+            return option::none()
+        };
+
+        option::some(next_iter_or_die(tree, iter))
+    }
+
+    /// Returns the next iterator, aborts if already at the end iterator.
+    /// Requires the tree is not changed after the input iterator is generated.
+    public fun next_iter_or_die<V>(tree: &SmartTree<V>, iter: Iterator): Iterator {
+        let node_index = iter.node_index;
+
+        let node = bucket_table::borrow(&tree.nodes, node_index);
```

Comment:
Does it abort here if it's the end iterator? Could we make it explicit like `assert!(node_index != NULL_INDEX, SOME_ERROR_CODE)` or something like this?

Reply:
Done